server/service/src/invoice/inbound_return/update/generate.rs

```diff
@@ -0,0 +1,194 @@
+}
+
+pub(crate) fn generate(
+    connection: &StorageConnection,
+    user_id: &str,
+    existing_return: InvoiceRow,
+    patch: UpdateInboundReturn,
+) -> Result<GenerateResult, UpdateInboundReturnError> {
+    let mut updated_return = existing_return.clone();
+
+    updated_return.user_id = Some(user_id.to_string());
+    updated_return.comment = patch.comment.clone().or(updated_return.comment);
+    updated_return.on_hold = patch.on_hold.unwrap_or(updated_return.on_hold);
+    updated_return.colour = patch.colour.clone().or(updated_return.colour);
+
+    set_new_status_datetime(&mut updated_return, &patch);
+
+    if let Some(status) = patch.status.clone() {
+        updated_return.status = status.as_invoice_row_status()
+    }
+
+    let should_create_batches = should_create_batches(&existing_return.status, &patch);
+
+    let batches_to_update = if should_create_batches {
+        Some(generate_lines_and_stock_lines(
+            connection,
+            &updated_return.store_id,
+            &updated_return.id,
+            &updated_return.name_link_id,
+        )?)
+    } else {
+        None
+    };
+
+    Ok(GenerateResult {
+        batches_to_update,
+        updated_return,
+    })
+}
+
+fn changed_status(
+    status: Option<UpdateInboundReturnStatus>,
+    existing_status: &InvoiceRowStatus,
+) -> Option<UpdateInboundReturnStatus> {
+    let new_status = match status {
+        Some(status) => status,
+        None => return None, // Status is not changing
+    };
+
+    if &new_status.as_invoice_row_status() == existing_status {
+        // The invoice already has this status, there's nothing to do.
+        return None;
+    }
+
+    Some(new_status)
+}
+
+pub fn should_create_batches(
+    existing_status: &InvoiceRowStatus,
+    patch: &UpdateInboundReturn,
+) -> bool {
+    let new_status = match changed_status(patch.status.to_owned(), existing_status) {
+        Some(status) => status,
+        None => return false, // There's no status to update
+    };
+
+    match (existing_status, new_status) {
+        (
+            // From New/Picked/Shipped to Delivered
+            InvoiceRowStatus::New | InvoiceRowStatus::Picked | InvoiceRowStatus::Shipped,
+            UpdateInboundReturnStatus::Delivered,
+        ) => true,
+        _ => false,
+    }
+}
+
+fn set_new_status_datetime(inbound_return: &mut InvoiceRow, patch: &UpdateInboundReturn) {
+    let new_status = match changed_status(patch.status.to_owned(), &inbound_return.status) {
+        Some(status) => status,
+        None => return, // There's no status to update
+    };
+
+    let current_datetime = Utc::now().naive_utc();
+
+    match (&inbound_return.status, new_status) {
+        // From New/Picked/Shipped to Delivered
+        (
+            InvoiceRowStatus::New | InvoiceRowStatus::Picked | InvoiceRowStatus::Shipped,
+            UpdateInboundReturnStatus::Delivered,
+        ) => {
+            inbound_return.delivered_datetime = Some(current_datetime);
```

Comment:
In the above cases, should picked/shipped datetime also get set? Or they can just stay None? I've seen both...

Reply:
hm. maybe that depends who did it? not ideal to have both.
I would say 'None' and have them updated only when explicitly set to that status value